src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2412,101 +2414,110 @@ public override BoundNode VisitConditionalOperator(BoundConditionalOperator node
-                BoundExpression alternativePlaceholder = CreatePlaceholderIfNecessary(alternative, alternativeResult);
+                BoundExpression consequencePlaceholder = CreatePlaceholderIfNecessary(consequence, consequenceRValue.ToTypeWithAnnotations());
+                BoundExpression alternativePlaceholder = CreatePlaceholderIfNecessary(alternative, alternativeRValue.ToTypeWithAnnotations());
                 HashSet<DiagnosticInfo> useSiteDiagnostics = null;
-                // https://github.com/dotnet/roslyn/issues/30432: InferBestTypeForConditionalOperator should use node.IsRef.
                 resultType = BestTypeInferrer.InferBestTypeForConditionalOperator(consequencePlaceholder, alternativePlaceholder, _conversions, out _, ref useSiteDiagnostics);
             }
 
-            TypeWithAnnotations visitResult;
+            TypeWithAnnotations lValueResult;
+            TypeWithState rValueResult;
             if ((object)resultType != null)
             {
-                var resultTypeWithAnnotations = TypeWithAnnotations.Create(resultType);
-                TypeWithAnnotations convertedConsequenceResult = default;
-                TypeWithAnnotations convertedAlternativeResult = default;
-
-                if (!isConstantFalse)
-                {
-                    convertedConsequenceResult = convertResult(
-                        node.Consequence,
-                        consequence,
-                        consequenceConversion,
-                        resultTypeWithAnnotations,
-                        consequenceResult);
-                }
-
-                if (!isConstantTrue)
+                if (isByRef)
                 {
-                    convertedAlternativeResult = convertResult(
-                        node.Alternative,
-                        alternative,
-                        alternativeConversion,
-                        resultTypeWithAnnotations,
-                        alternativeResult);
-                }
+                    if (IsNullabilityMismatch(consequenceLValue, alternativeLValue))
+                    {
+                        // l-value types must match
+                        ReportNullabilityMismatchInAssignment(node.Syntax, consequenceLValue, alternativeLValue);
+                    }
 
-                if (!convertedAlternativeResult.HasType)
-                {
-                    Debug.Assert(convertedConsequenceResult.HasType);
-                    visitResult = convertedConsequenceResult;
-                }
-                else if (!convertedConsequenceResult.HasType)
-                {
-                    Debug.Assert(convertedAlternativeResult.HasType);
-                    visitResult = convertedAlternativeResult;
+                    lValueResult = TypeWithAnnotations.Create(resultType, consequenceLValue.NullableAnnotation.Join(alternativeLValue.NullableAnnotation));
```

Comment:
>Join [](start = 111, length = 4)

This should be `EnsureCompatible` rather than `Join` (refs are not variant)

Reply:
Good catch. Added a test to illustrate.

---
In reply to: [268289616](https://github.com/dotnet/roslyn/pull/34323#discussion_r268289616) [](ancestors = 268289616)